src/main.c

```diff
@@ -212,21 +210,33 @@ static void dump_test_signature(elf_t *elf)
         fprintf(f, "%08x\n", memory_read_w(addr));
 
     fclose(f);
+    elf_delete(elf);
 }
 
+#define MEM_SIZE 0xFFFFFFFFULL  /* 2^32 - 1 */
+#define STACK_SIZE 0x1000       /* 4096 */
+#define ARGS_OFFSET_SIZE 0x1000 /* 4096 */
+
 int main(int argc, char **args)
 {
     if (argc == 1 || !parse_args(argc, args)) {
         print_usage(args[0]);
         return 1;
     }
 
-    /* open the ELF file from the file system */
-    elf_t *elf = elf_new();
-    if (!elf_open(elf, opt_prog_name)) {
-        fprintf(stderr, "Unable to open ELF file '%s'\n", opt_prog_name);
-        return 1;
-    }
+    vm_attr_t attr = {
+        .mem_size = MEM_SIZE,
+        .stack_size = STACK_SIZE,
+        .args_offset_size = ARGS_OFFSET_SIZE,
+        .argc = prog_argc,
+        .argv = prog_args,
+        .logging_level = 0,
+        .emu_data.vm_user = malloc(sizeof(vm_user_t)),
+        .cycle_per_step = 100,
+        .allow_misalign = opt_misaligned,
+    };
```

Comment:
Since we are going to make these parameters user-defined now, I will suggest having some checks (`check_vm_attr()`) on the `vm_attr_t` before the emulator creation stage, otherwise bad parameters can lead to a crash which is not easy to figure out quickly. For example:  `mem_size` should be larger than `stack_size + .args_offset_size`.

Note that `mem_size` should also be larger than the ELF mapping range to load correctly, but this may not be easy to implement at `check_vm_attr()`. As an alternative, maybe apply the range check for `memory_write` is a simple but great idea.

Reply:
Nice perspective. I think validation of `vm_attr_t` could be complex. For example, shall we check the invalid file during validation or propagate to `elf_open` to verify the invalidation. Similar situation for memory, if the user-provided memory size is not enough to run emulation, shall we simply abort before `rv_create` or propagate to segmentation fault. We may need more discussion about this. Maybe in future PR?

Reply:
Sure. We can leave these at other place.

Reply:
Sure. We can leave these at other places.

Reply:
> Sure. We can leave these at other places.

I have mentioned it in the commit message.